app/src/main/java/org/stepic/droid/ui/adapters/CodeToolbarAdapter.kt

```diff
@@ -28,16 +57,43 @@ class CodeToolbarAdapter(private val context: Context) : RecyclerView.Adapter<Co
         notifyDataSetChanged()
     }
 
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CodeToolbarItem? {
-        val view = LayoutInflater.from(parent.context).inflate(R.layout.view_code_toolbar_item, parent, false)
-        return CodeToolbarItem(view, onItemClickListener)
+    override fun getItemViewType(position: Int) =
+            if (autoCompleteWords.isNotEmpty() && position == autoCompleteWords.size) {
+                SEPARATOR_VIEW_TYPE
+            } else {
+                ELEMENT_VIEW_TYPE
+            }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder? {
+        val inflater = LayoutInflater.from(parent.context)
+        return when (viewType) {
+            SEPARATOR_VIEW_TYPE ->
+                CodeToolbarSeparator(inflater.inflate(R.layout.view_code_toolbar_separator, parent, false))
+
+            ELEMENT_VIEW_TYPE ->
+                CodeToolbarItem(inflater.inflate(R.layout.view_code_toolbar_item, parent, false), onItemClickListener)
+
+            else -> throw IllegalArgumentException("Wrong view type: $viewType")
+        }
     }
 
-    override fun onBindViewHolder(holder: CodeToolbarItem, position: Int) {
-        holder.bindData(symbols[position])
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        when (getItemViewType(position)) {
+            ELEMENT_VIEW_TYPE -> (holder as CodeToolbarItem).let {
+                if (position < autocompleteItemsCount) {
+                    val word = SpannableString(autoCompleteWords[position])
+                    autoCompletePrefix?.let {
+                        word.setSpan(autocompletePrefixBackgroundSpan, 0, it.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+                    }
+                    it.bindData(word)
+                } else {
+                    it.bindData(SpannableString(symbols[position - autocompleteItemsCount]))
+                }
```

Comment:
we create a lot of `SpannableString` objects in `onBindViewHolder`. It is executed, when user scrolls toolbar

Reply:
Let's use `var autoCompleteWords: List<SpannableString>` with included prefix